zebra-state/src/service/check/tests/utxo.rs

```diff
@@ -0,0 +1,764 @@
+        mut prevout_input in TypeNameToDebug::<transparent::Input>::arbitrary_with(None),
+        use_finalized_state in any::<bool>(),
+    ) {
+        zebra_test::init();
+
+        let mut block1 = zebra_test::vectors::BLOCK_MAINNET_1_BYTES
+            .zcash_deserialize_into::<Block>()
+            .expect("block should deserialize");
+
+        // create an output
+        let output_transaction = transaction_v4_with_transparent_data([], [output.0]);
+
+        // create a spend
+        let expected_outpoint = transparent::OutPoint { hash: output_transaction.hash(), index: 0 };
+        prevout_input.set_outpoint(expected_outpoint);
+        let spend_transaction = transaction_v4_with_transparent_data([prevout_input.0], []);
+
+        // convert the coinbase transaction to a version that the non-finalized state will accept
+        block1.transactions[0] = transaction_v4_from_coinbase(&block1.transactions[0]).into();
+
+        block1
+            .transactions
+            .extend([output_transaction.into(), spend_transaction.into()]);
+
+        let (mut state, _genesis) = new_state_with_mainnet_genesis();
+        let previous_mem = state.mem.clone();
+
+        // randomly choose to commit the block to the finalized or non-finalized state
+        if use_finalized_state {
+            let block1 = FinalizedBlock::from(Arc::new(block1));
+            let commit_result = state.disk.commit_finalized_direct(block1.clone(), "test");
+
+            // the block was committed
+            prop_assert_eq!(Some((Height(1), block1.hash)), state.best_tip());
+            prop_assert!(commit_result.is_ok());
+
+            // the non-finalized state didn't change
+            prop_assert!(state.mem.eq_internal_state(&previous_mem));
+
+            // the finalized state added then spent the UTXO
+            prop_assert!(state.disk.utxo(&expected_outpoint).is_none());
+            // the non-finalized state does not have the UTXO
+            prop_assert!(state.mem.any_utxo(&expected_outpoint).is_none());
+        } else {
+            let block1 = Arc::new(block1).prepare();
+            let commit_result =
+                state.validate_and_commit(block1.clone());
+
+            // the block was committed
+            prop_assert_eq!(commit_result, Ok(()));
+            prop_assert_eq!(Some((Height(1), block1.hash)), state.best_tip());
+
+            // the block data is in the non-finalized state
+            prop_assert!(!state.mem.eq_internal_state(&previous_mem));
+
+            // the non-finalized state has the spent its own UTXO
+            prop_assert_eq!(state.mem.chain_set.len(), 1);
+            prop_assert!(!state.mem.chain_set.iter().next().unwrap().unspent_utxos().contains_key(&expected_outpoint));
+            prop_assert!(state.mem.chain_set.iter().next().unwrap().created_utxos.contains_key(&expected_outpoint));
+            prop_assert!(state.mem.chain_set.iter().next().unwrap().spent_utxos.contains(&expected_outpoint));
```

Comment:
Suggestion: maybe use a binding to reduce the repeated code?

```suggestion
            prop_assert_eq!(state.mem.chain_set.len(), 1);
            let chain = state.mem.chain_set.iter().next().unwrap();
            prop_assert!(!chain.unspent_utxos().contains_key(&expected_outpoint));
            prop_assert!(chain.created_utxos.contains_key(&expected_outpoint));
            prop_assert!(chain.spent_utxos.contains(&expected_outpoint));
```

Reply:
Similarly, see 0f86cd49 in PR #2515.